src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -415,15 +502,231 @@ private void CheckDisposedOrStarted()
+            }
+        }
+
+        private SafeCurlSlistHandle SetRequestHeaders(SafeCurlHandle handle, HttpRequestMessage request)
+        {
+            SafeCurlSlistHandle retVal = new SafeCurlSlistHandle();
+            if (request.Headers == null)
+            {
+                return retVal;
+            }
+
+            HttpHeaders contentHeaders = null;
+            if (request.Content != null)
+            {
+                // TODO: Content-Length header isn't getting correctly placed using ToString()
+                // This is a bug in HttpContentHeaders that needs to be fixed.
+                if (request.Content.Headers.ContentLength.HasValue)
+                {
+                    long contentLength = request.Content.Headers.ContentLength.Value;
+                    request.Content.Headers.ContentLength = null;
+                    request.Content.Headers.ContentLength = contentLength;
+                }
+                contentHeaders = request.Content.Headers;
+            }
+
+            string[] allHeaders = HeaderUtilities.DumpHeaders(request.Headers, contentHeaders)
+                .Split(s_headerDelimiters, StringSplitOptions.RemoveEmptyEntries);
+            bool gotReference = false;
+            try
+            {
+                for (int i = 0; i < allHeaders.Length; i++)
+                {
+                    string header = allHeaders[i];
+                    if (header.Equals("{") || header.Equals("}"))
+                    {
+                        continue;
+                    }
+                    IntPtr rawHandle = IntPtr.Zero;
+                    if (!retVal.IsInvalid)
+                    {
+                        retVal.DangerousAddRef(ref gotReference);
+                        rawHandle = retVal.DangerousGetHandle();
+                    }
+                    rawHandle = Interop.libcurl.curl_slist_append(rawHandle, header);
+                    retVal.SetHandle(rawHandle);
+                }
+            }
+            finally
+            {
+                if (gotReference)
+                {
+                    retVal.DangerousRelease();
+                }
+            }
+
+            if (!retVal.IsInvalid)
+            {
+                try
+                {
+                    retVal.DangerousAddRef(ref gotReference);
+                    SetCurlOption(handle, CURLoption.CURLOPT_HTTPHEADER, retVal.DangerousGetHandle());
+                }
+                finally
+                {
+                    if (gotReference)
+                    {
+                        retVal.DangerousRelease();
+                    }
+                }
+            }
+
+            return retVal;
+        }
+
+        private void AddEasyHandle(RequestCompletionSource state)
+        {
+            bool gotReference = false;
+            SafeCurlHandle requestHandle = state.RequestHandle;
+            try
+            {
+                requestHandle.DangerousAddRef(ref gotReference);
```

Comment:
(That said, this pattern makes me question whether a SafeHandle is really appropriate here.  If for the majority of the lifetime of the handle we've marked it as DangerousAddRef, that's a very long window where something could go wrong and we drop the safe handle such that the handle will never get closed and never be finalized.  At that point, the safe handle isn't buying us much over just storing an IntPtr.  It's fine for now, and maybe there are other benefits I'm not seeing clearly at the moment, but something to think about.)